packages/application-shell/src/components/quick-access/butler/butler.tsx

```diff
@@ -0,0 +1,742 @@
+      // happens to be where the results will pop up
+      skipNextSelection.current = true;
+
+      if (event.key === 'ArrowDown') {
+        // prevent cursor from jumping to end of text input
+        event.preventDefault();
+        dispatch({ type: 'incrementSelectedResult' });
+        return;
+      }
+      if (event.key === 'ArrowUp') {
+        // browse through history
+        if (
+          state.searchText.length === 0 ||
+          (state.selectedResult < 1 && state.enableHistory)
+        ) {
+          shouldSelectFieldText.current = true;
+          const selectedIndex =
+            state.searchText.length === 0
+              ? // When going back the first step
+                -1
+              : // When going back more than one step
+                props.historyEntries.findIndex(
+                  command => command.searchText === state.searchText
+                );
+          // Pick the previous command from the history
+          const prevCommand =
+            selectedIndex === -1
+              ? // previous command on top of the history when going back on
+                // first step
+                last(props.historyEntries)
+              : // previous command is deeper down
+                // When the history does not exist (negative index), then
+                // this implicitly returns undefined
+                props.historyEntries[selectedIndex - 1];
+          // Skip when no previous entry exists in the history
+          if (!prevCommand) return;
+          dispatch({
+            type: 'pickCommandFromHistory',
+            payload: {
+              searchText: prevCommand.searchText,
+              results: prevCommand.results,
+            },
+          });
+          return;
+        }
+        // prevent cursor from jumping to beginning of text input
+        event.preventDefault();
+        dispatch({ type: 'decrementSelectedResult' });
+        return;
+      }
+      if (state.selectedResult > -1) {
+        if (event.key === 'ArrowRight') {
+          const command = state.results[state.selectedResult];
+          const searchText = state.searchText;
+          const isCursorAtEnd =
+            searchInputRef.current &&
+            state.searchText.length === searchInputRef.current.selectionStart;
+
+          const isEverythingSelected =
+            searchInputRef.current &&
+            searchInputRef.current.selectionStart === 0 &&
+            state.searchText.length === searchInputRef.current.selectionEnd;
+
+          // only allow diving in when cursor is at end of input or when
+          // the complete text is selected (when browsing through history)
+          if (!isCursorAtEnd && !isEverythingSelected) return;
+
+          unsetHasNetworkError();
+
+          // NOTE: since we need to fetch the "next command", which is an async operation,
+          // we use a IIFE to process that and eventually update the state.
+          (async () => {
```

Comment:
This was one of the edge cases. Getting the "next" commands returns a promise, because some of the subcommands require to make a network request to fetch some data. As a consequence we need to treat this asynchronously.

Before I think it was kind of broken, or at least the promises were not properly "awaited".